### DISCUSSION TITLE
Replacement for ItemEngine + core_props behavior in v0.12.x?

### DISCUSSION THREAD

**Discussion by demery** [LeMyst/WikibaseIntegrator#437] [Q&A]
Hi Everyone, 
Apologies if this is a redundant post. I've looked but haven't found an answer to my question. We've inherited code based on v0.11.x that makes heavy use of the `ItemEngine` + `core_props` method to locate items in our Wikibase instance by property value. 

Is there a replacement for locating items by non-`entity_id` property values in v0.12.x? Or, should this be done via SPARQL query now? V0.12.x has certain features we now need, and I'm doing research on the level of effort needed to update the code.

With much thanks for your help,
Doug

**Comment by LeMyst** [comment]
Hello @demery 

Do you have a code snippet to share ? To see how you use core_props ?

**Reply by demery** [reply]
Happily. 

One example in several: We store Wikidata IDs with the names of people, organizations, and languages in our Wikibase using property `P48`. We have process that links those names to other records. To find the Wikibase Item and `item_id` we use `ItemEngine()` thus:

```python
        actor_statements = []
        wd_qid = 'Q868' # Aristotle
        # Wikidata QID (P48)
        actor_statements += [wbi_datatype.String(value=wd_qid, prop_nr='P48')]
        actor_item = wbi_core.ItemEngine(data=actor_statements, core_props=['P48'], search_only=True)
```

In some contexts, we use ItemEngine in a "find or create mode" to create or locate or add an item to our WB instance:

```python
    # no name exists; create the item
    actor_statements = []
    wd_qid = actor_data['wikidata_qid'] # Aristotle
    dsname_id = IDService.generate_new_id() 

    # Wikidata QID (P48)
    actor_statements += [wbi_datatype.String(value=wd_qid, prop_nr='P48')]
    # DS ID (P1)
    actor_statements += [wbi_datatype.String(value=dsname_id, prop_nr='P1')]

   # actor_data is a dict of actor values
    instance_of = actor_data['instance_of']
    # instance of (P42)
    if instance_of:
        if instance_of.lower().strip() == 'human':
            # Personal (Name Authority) (Q18)
            actor_statements += [wbi_datatype.ItemID(value='Q18', prop_nr='P42', if_exists='APPEND')]
        if instance_of.lower().strip() == 'corporate':
            # Corporate (Name Authority) (Q19)
            actor_statements += [wbi_datatype.ItemID(value='Q19', prop_nr='P42', if_exists='APPEND')]

    wb_item = wbi_core.ItemEngine(data=actor_statements, core_props=['P48'])
```

**Reply by LeMyst** [reply]
I don't have a solution for you at this moment.

I think what comes closest is the wd-population script I made:
https://github.com/LeMyst/wd-population/blob/master/departements.py
that work with the work-in-progress https://github.com/LeMyst/WikibaseIntegrator/tree/new-fastrun branch of WikibaseIntegrator.

there is faster version (but non functional right now):
https://github.com/LeMyst/wd-population/blob/master/departements_fastrun.py

The wd-population script search and update if found.

I'll try to fix the new fastrun branch this weekend.

I'll keep you in touch.

**Reply by demery** [reply]
Thanks. I'll check this out and look for updates on the branch.
